packages/ramps-controller/src/RampsController.ts

```diff
@@ -0,0 +1,233 @@
+  handler: RampsController['getCountries'];
+};
+
+export type RampsControllerActions =
+  | RampsControllerGetStateAction
+  | RampsControllerGetCountriesAction;
+
+export type RampsControllerStateChangeEvent = ControllerStateChangeEvent<
+  typeof controllerName,
+  RampsControllerState
+>;
+
+export type RampsControllerEvents = RampsControllerStateChangeEvent;
+
+export type RampsControllerMessenger = Messenger<
+  typeof controllerName,
+  RampsControllerActions,
+  RampsControllerEvents
+>;
+
+const rampsControllerMetadata = {
+  metamaskEnvironment: {
+    persist: true,
+    anonymous: true,
+    includeInDebugSnapshot: true,
+    includeInStateLogs: true,
+    usedInUi: true,
+  },
+  context: {
+    persist: true,
+    anonymous: true,
+    includeInDebugSnapshot: true,
+    includeInStateLogs: true,
+    usedInUi: true,
+  },
+  region: {
+    persist: true,
+    anonymous: true,
+    includeInDebugSnapshot: true,
+    includeInStateLogs: true,
+    usedInUi: true,
+  },
+};
+
+const defaultState: RampsControllerState = {
+  metamaskEnvironment: SdkEnvironment.Staging,
+  context: Context.Browser,
+  region: null,
+};
+
+enum ApiService {
+  Regions = 'regions',
+}
+
+/**
+ * Controller that manages on-ramp and off-ramp operations.
+ * The ramps controller is responsible for handling cryptocurrency purchase and sale operations.
+ *
+ */
+export class RampsController extends BaseController<
+  typeof controllerName,
+  RampsControllerState,
+  RampsControllerMessenger
+> {
+  /**
+   * Constructor for RampsController.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The messenger object.
+   * @param options.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: RampsControllerMessenger;
+    state: Partial<RampsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: rampsControllerMetadata,
+      state: { ...defaultState, ...state } as RampsControllerState,
+    });
+
+    this.#registerMessageHandlers();
+  }
+
+  /**
+   * Gets the non-cached API URL based on the metamask environment.
+   *
+   * @param service - The API service to append to the base URL.
+   * @returns The non-cached API URL based on the metamask environment.
+   */
+  #getApiUrl(service?: ApiService): string {
+    let url = 'http://localhost:3000';
+    if (this.state.metamaskEnvironment === SdkEnvironment.Production) {
+      url = 'https://on-ramp.api.cx.metamask.io';
+    } else if (this.state.metamaskEnvironment === SdkEnvironment.Staging) {
+      url = 'https://on-ramp.uat-api.cx.metamask.io';
+    }
+
+    const urlWithPath = new URL(service ?? '', url);
+    return urlWithPath.toString();
+  }
+
+  async #getGeolocation(): Promise<string> {
+    try {
+      const url = this.#getApiUrl();
+      const response = await fetch(new URL('geolocation', url).toString());
```

Comment:
Instead of making HTTP requests in your controller, we recommend extracting an API client from your controller that represents the "on ramp" API. It could serve as a direct layer on top of the layer and could be used independent of the controller anywhere in the stack. 

We have a pattern called "data services" that would be useful here. Here is a tutorial we wrote that shows you how to make one: https://github.com/MetaMask/core/blob/main/docs/data-services.md. In addition we have an example data service here: https://github.com/MetaMask/core/blob/main/packages/sample-controllers/src/sample-gas-prices-service/sample-gas-prices-service.ts.

Let me know your thoughts on this.